File: fontes/frontend/fenix_provider/lib/src/model/financeiro/fin_parcela_pagar.dart
/*
Title: T2Ti ERP 3.0                                                                
Description: Model relacionado à tabela [FIN_PARCELA_PAGAR] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';

import 'package:intl/intl.dart';
import 'package:fenix/src/model/model.dart';

class FinParcelaPagar {
	int? id;
	int? idFinLancamentoPagar;
	int? idFinStatusParcela;
	int? idFinTipoPagamento;
	int? idFinChequeEmitido;
	int? numeroParcela;
	DateTime? dataEmissao;
	DateTime? dataVencimento;
	DateTime? dataPagamento;
	DateTime? descontoAte;
	double? valor;
	double? taxaJuro;
	double? taxaMulta;
	double? taxaDesconto;
	double? valorJuro;
	double? valorMulta;
	double? valorDesconto;
	double? valorPago;
	String? historico;
	FinStatusParcela? finStatusParcela;
	FinTipoPagamento? finTipoPagamento;
	FinChequeEmitido? finChequeEmitido;
	FinLancamentoPagar? finLancamentoPagar;

	FinParcelaPagar({
			id,
			idFinLancamentoPagar,
			idFinStatusParcela,
			idFinTipoPagamento,
			idFinChequeEmitido,
			numeroParcela,
			dataEmissao,
			dataVencimento,
			dataPagamento,
			descontoAte,
			valor,
			taxaJuro,
			taxaMulta,
			taxaDesconto,
			valorJuro,
			valorMulta,
			valorDesconto,
			valorPago,
			historico,
			finStatusParcela,
			finTipoPagamento,
			finChequeEmitido,
			finLancamentoPagar,
		});

	static List<String> campos = <String>[
		'ID', 
		'NUMERO_PARCELA', 
		'DATA_EMISSAO', 
		'DATA_VENCIMENTO', 
		'DATA_PAGAMENTO', 
		'DESCONTO_ATE', 
		'VALOR', 
		'TAXA_JURO', 
		'TAXA_MULTA', 
		'TAXA_DESCONTO', 
		'VALOR_JURO', 
		'VALOR_MULTA', 
		'VALOR_DESCONTO', 
		'VALOR_PAGO', 
		'HISTORICO', 
	];
	
	static List<String> colunas = <String>[
		'Id', 
		'Número da Parcela', 
		'Data de Emissão', 
		'Data de Vencimento', 
		'Data de Pagamento', 
		'Desconto Ate', 
		'Valor', 
		'Taxa Juros', 
		'Taxa Multa', 
		'Taxa Desconto', 
		'Valor Juros', 
		'Valor Multa', 
		'Valor Desconto', 
		'Valor Pago', 
		'Histórico', 
	];

	FinParcelaPagar.fromJson(Map<String, dynamic> jsonDados) {
		id = jsonDados['id'];
		idFinLancamentoPagar = jsonDados['idFinLancamentoPagar'];
		idFinStatusParcela = jsonDados['idFinStatusParcela'];
		idFinTipoPagamento = jsonDados['idFinTipoPagamento'];
		idFinChequeEmitido = jsonDados['idFinChequeEmitido'];
		numeroParcela = jsonDados['numeroParcela'];
		dataEmissao = jsonDados['dataEmissao'] != null ? DateTime.tryParse(jsonDados['dataEmissao']) : null;
		dataVencimento = jsonDados['dataVencimento'] != null ? DateTime.tryParse(jsonDados['dataVencimento']) : null;
		dataPagamento = jsonDados['dataPagamento'] != null ? DateTime.tryParse(jsonDados['dataPagamento']) : null;
		descontoAte = jsonDados['descontoAte'] != null ? DateTime.tryParse(jsonDados['descontoAte']) : null;
		valor = jsonDados['valor']?.toDouble();
		taxaJuro = jsonDados['taxaJuro']?.toDouble();
		taxaMulta = jsonDados['taxaMulta']?.toDouble();
		taxaDesconto = jsonDados['taxaDesconto']?.toDouble();
		valorJuro = jsonDados['valorJuro']?.toDouble();
		valorMulta = jsonDados['valorMulta']?.toDouble();
		valorDesconto = jsonDados['valorDesconto']?.toDouble();
		valorPago = jsonDados['valorPago']?.toDouble();
		historico = jsonDados['historico'];
		finStatusParcela = jsonDados['finStatusParcela'] == null ? null : FinStatusParcela.fromJson(jsonDados['finStatusParcela']);
		finTipoPagamento = jsonDados['finTipoPagamento'] == null ? null : FinTipoPagamento.fromJson(jsonDados['finTipoPagamento']);
		finChequeEmitido = jsonDados['finChequeEmitido'] == null ? null : FinChequeEmitido.fromJson(jsonDados['finChequeEmitido']);
		finLancamentoPagar = jsonDados['finLancamentoPagar'] == null ? null : FinLancamentoPagar.fromJson(jsonDados['finLancamentoPagar']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonDados = <String, dynamic>{};

		jsonDados['id'] = id ?? 0;
		jsonDados['idFinLancamentoPagar'] = idFinLancamentoPagar ?? 0;
		jsonDados['idFinStatusParcela'] = idFinStatusParcela ?? 0;
		jsonDados['idFinTipoPagamento'] = idFinTipoPagamento ?? 0;
		jsonDados['idFinChequeEmitido'] = idFinChequeEmitido ?? 0;
		jsonDados['numeroParcela'] = numeroParcela ?? 0;
		jsonDados['dataEmissao'] = dataEmissao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissao!) : null;
		jsonDados['dataVencimento'] = dataVencimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataVencimento!) : null;
		jsonDados['dataPagamento'] = dataPagamento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataPagamento!) : null;
		jsonDados['descontoAte'] = descontoAte != null ? DateFormat('yyyy-MM-ddT00:00:00').format(descontoAte!) : null;
		jsonDados['valor'] = valor;
		jsonDados['taxaJuro'] = taxaJuro;
		jsonDados['taxaMulta'] = taxaMulta;
		jsonDados['taxaDesconto'] = taxaDesconto;
		jsonDados['valorJuro'] = valorJuro;
		jsonDados['valorMulta'] = valorMulta;
		jsonDados['valorDesconto'] = valorDesconto;
		jsonDados['valorPago'] = valorPago;
		jsonDados['historico'] = historico;
		jsonDados['finStatusParcela'] = finStatusParcela == null ? null : finStatusParcela!.toJson;
		jsonDados['finTipoPagamento'] = finTipoPagamento == null ? null : finTipoPagamento!.toJson;
		jsonDados['finChequeEmitido'] = finChequeEmitido == null ? null : finChequeEmitido!.toJson;
		jsonDados['finLancamentoPagar'] = finLancamentoPagar == null ? null : finLancamentoPagar!.toJson;
	
		return jsonDados;
	}
	

	String objetoEncodeJson(FinParcelaPagar objeto) {
	  final jsonDados = objeto.toJson;
	  return json.encode(jsonDados);
	}
	
}

File: fontes/backend/php/fenix-doctrine/src/model/financeiro/FinParcelaPagar.php
<?php
/*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Model relacionado à tabela [FIN_PARCELA_PAGAR] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
declare(strict_types=1);

use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 * @ORM\Table(name="FIN_PARCELA_PAGAR")
 */
class FinParcelaPagar extends ModelBase implements \JsonSerializable
{
	/**
	 * @ORM\Id
	 * @ORM\Column(type="integer")
	 * @ORM\GeneratedValue(strategy="AUTO")
	 */
	private $id;

	/**
	 * @ORM\Column(type="integer", name="ID_FIN_LANCAMENTO_PAGAR")
	 */
	private $idFinLancamentoPagar;
	
	/**
	 * @ORM\Column(type="integer", name="NUMERO_PARCELA")
	 */
	private $numeroParcela;

	/**
	 * @ORM\Column(type="date", name="DATA_EMISSAO")
	 */
	private $dataEmissao;

	/**
	 * @ORM\Column(type="date", name="DATA_VENCIMENTO")
	 */
	private $dataVencimento;

	/**
	 * @ORM\Column(type="date", name="DATA_PAGAMENTO")
	 */
	private $dataPagamento;
	
	/**
	 * @ORM\Column(type="date", name="DESCONTO_ATE")
	 */
	private $descontoAte;

	/**
	 * @ORM\Column(type="float", name="VALOR")
	 */
	private $valor;

	/**
	 * @ORM\Column(type="float", name="TAXA_JURO")
	 */
	private $taxaJuro;

	/**
	 * @ORM\Column(type="float", name="TAXA_MULTA")
	 */
	private $taxaMulta;

	/**
	 * @ORM\Column(type="float", name="TAXA_DESCONTO")
	 */
	private $taxaDesconto;

	/**
	 * @ORM\Column(type="float", name="VALOR_JURO")
	 */
	private $valorJuro;

	/**
	 * @ORM\Column(type="float", name="VALOR_MULTA")
	 */
	private $valorMulta;

	/**
	 * @ORM\Column(type="float", name="VALOR_DESCONTO")
	 */
	private $valorDesconto;

	/**
	 * @ORM\Column(type="float", name="VALOR_PAGO")
	 */
	private $valorPago;

	/**
	 * @ORM\Column(type="string", name="HISTORICO")
	 */
	private $historico;


    /**
     * Relations
     */
    
    /**
     * @ORM\OneToOne(targetEntity="FinStatusParcela", fetch="EAGER")
     * @ORM\JoinColumn(name="ID_FIN_STATUS_PARCELA", referencedColumnName="id")
     */
    private $finStatusParcela;

    /**
     * @ORM\OneToOne(targetEntity="FinTipoPagamento", fetch="EAGER")
     * @ORM\JoinColumn(name="ID_FIN_TIPO_PAGAMENTO", referencedColumnName="id")
     */
    private $finTipoPagamento;

    /**
     * @ORM\OneToOne(targetEntity="FinChequeEmitido", fetch="EAGER")
     * @ORM\JoinColumn(name="ID_FIN_CHEQUE_EMITIDO", referencedColumnName="id")
     */
    private $finChequeEmitido;

    /**
     * @ORM\ManyToOne(targetEntity="FinLancamentoPagar", inversedBy="listaFinParcelaPagar")
     * @ORM\JoinColumn(name="ID_FIN_LANCAMENTO_PAGAR", referencedColumnName="id")
     */
    private $finLancamentoPagar;


    /**
     * Gets e Sets
     */

    public function getId() 
	{
		return $this->id;
	}

	public function setId($id) 
	{
		$this->id = $id;
	}

    public function getIdFinLancamentoPagar() 
	{
		return $this->idFinLancamentoPagar;
	}

	public function setIdFinLancamentoPagar($idFinLancamentoPagar) 
	{
		$this->idFinLancamentoPagar = $idFinLancamentoPagar;
	}
	
    public function getNumeroParcela() 
	{
		return $this->numeroParcela;
	}

	public function setNumeroParcela($numeroParcela) 
	{
		$this->numeroParcela = $numeroParcela;
	}

    public function getDataEmissao() 
	{
		if ($this->dataEmissao != null) {
			return $this->dataEmissao->format('Y-m-d');
		} else {
			return null;
		}
	}
	public function setDataEmissao($dataEmissao) 
	{
		$this->dataEmissao = $dataEmissao != null ? new \DateTime($dataEmissao) : null;
	}

    public function getDataVencimento() 
	{
		if ($this->dataVencimento != null) {
			return $this->dataVencimento->format('Y-m-d');
		} else {
			return null;
		}
	}
	public function setDataVencimento($dataVencimento) 
	{
		$this->dataVencimento = $dataVencimento != null ? new \DateTime($dataVencimento) : null;
	}

    public function getDataPagamento() 
	{
		if ($this->dataPagamento != null) {
			return $this->dataPagamento->format('Y-m-d');
		} else {
			return null;
		}
	}
	public function setDataPagamento($dataPagamento) 
	{
		$this->dataPagamento = $dataPagamento != null ? new \DateTime($dataPagamento) : null;
	}
	
    public function getDescontoAte() 
	{
		if ($this->descontoAte != null) {
			return $this->descontoAte->format('Y-m-d');
		} else {
			return null;
		}
	}
	public function setDescontoAte($descontoAte) 
	{
		$this->descontoAte = $descontoAte != null ? new \DateTime($descontoAte) : null;
	}

    public function getValor() 
	{
		return $this->valor;
	}

	public function setValor($valor) 
	{
		$this->valor = $valor;
	}

    public function getTaxaJuro() 
	{
		return $this->taxaJuro;
	}

	public function setTaxaJuro($taxaJuro) 
	{
		$this->taxaJuro = $taxaJuro;
	}

    public function getTaxaMulta() 
	{
		return $this->taxaMulta;
	}

	public function setTaxaMulta($taxaMulta) 
	{
		$this->taxaMulta = $taxaMulta;
	}

    public function getTaxaDesconto() 
	{
		return $this->taxaDesconto;
	}

	public function setTaxaDesconto($taxaDesconto) 
	{
		$this->taxaDesconto = $taxaDesconto;
	}

    public function getValorJuro() 
	{
		return $this->valorJuro;
	}

	public function setValorJuro($valorJuro) 
	{
		$this->valorJuro = $valorJuro;
	}

    public function getValorMulta() 
	{
		return $this->valorMulta;
	}

	public function setValorMulta($valorMulta) 
	{
		$this->valorMulta = $valorMulta;
	}

    public function getValorDesconto() 
	{
		return $this->valorDesconto;
	}

	public function setValorDesconto($valorDesconto) 
	{
		$this->valorDesconto = $valorDesconto;
	}

    public function getValorPago() 
	{
		return $this->valorPago;
	}

	public function setValorPago($valorPago) 
	{
		$this->valorPago = $valorPago;
	}

    public function getHistorico() 
	{
		return $this->historico;
	}

	public function setHistorico($historico) 
	{
		$this->historico = $historico;
	}

    public function getFinLancamentoPagar(): ?FinLancamentoPagar 
	{
		return $this->finLancamentoPagar;
	}

	public function setFinLancamentoPagar(?FinLancamentoPagar $finLancamentoPagar) 
	{
		$this->finLancamentoPagar = $finLancamentoPagar;
	}

    public function getFinStatusParcela(): ?FinStatusParcela 
	{
		return $this->finStatusParcela;
	}

	public function setFinStatusParcela(?FinStatusParcela $finStatusParcela) 
	{
		$this->finStatusParcela = $finStatusParcela;
	}

    public function getFinTipoPagamento(): ?FinTipoPagamento 
	{
		return $this->finTipoPagamento;
	}

	public function setFinTipoPagamento(?FinTipoPagamento $finTipoPagamento) 
	{
		$this->finTipoPagamento = $finTipoPagamento;
	}

    public function getFinChequeEmitido(): ?FinChequeEmitido 
	{
		return $this->finChequeEmitido;
	}

	public function setFinChequeEmitido(?FinChequeEmitido $finChequeEmitido) 
	{
		$this->finChequeEmitido = $finChequeEmitido;
	}


    /**
     * Constructor
     */
    public function __construct($objetoJson)
    {
        if (isset($objetoJson)) {
            isset($objetoJson->id) ? $this->setId($objetoJson->id) : $this->setId(null);
            $this->mapear($objetoJson);
        }

		
    }

    /**
     * Mapping
     */
    public function mapear($objeto)
    {
        if (isset($objeto)) {
			$this->setNumeroParcela($objeto->numeroParcela);
			$this->setDataEmissao($objeto->dataEmissao);
			$this->setDataVencimento($objeto->dataVencimento);
			$this->setDescontoAte($objeto->descontoAte);
			$this->setDataPagamento($objeto->dataPagamento);
			$this->setValor($objeto->valor);
			$this->setTaxaJuro($objeto->taxaJuro);
			$this->setTaxaMulta($objeto->taxaMulta);
			$this->setTaxaDesconto($objeto->taxaDesconto);
			$this->setValorJuro($objeto->valorJuro);
			$this->setValorMulta($objeto->valorMulta);
			$this->setValorDesconto($objeto->valorDesconto);
			$this->setValorPago($objeto->valorPago);
			$this->setHistorico($objeto->historico);
			$this->setIdFinLancamentoPagar($objeto->idFinLancamentoPagar);
		}
    }


    /**
     * Validation
     */
    public function validarObjetoRequisicao($objJson, $id)
    {
        return parent::validarObjeto($objJson, $id, 'FinParcelaPagar');
    }


    /**
     * Serialization
     * {@inheritdoc}
     */
    public function jsonSerialize()
    {
        return [
			'id' => $this->getId(),
			'idFinLancamentoPagar' => $this->getIdFinLancamentoPagar(),
			'numeroParcela' => $this->getNumeroParcela(),
			'dataEmissao' => $this->getDataEmissao(),
			'dataVencimento' => $this->getDataVencimento(),
			'dataPagamento' => $this->getDataPagamento(),
			'descontoAte' => $this->getDescontoAte(),
			'valor' => $this->getValor(),
			'taxaJuro' => $this->getTaxaJuro(),
			'taxaMulta' => $this->getTaxaMulta(),
			'taxaDesconto' => $this->getTaxaDesconto(),
			'valorJuro' => $this->getValorJuro(),
			'valorMulta' => $this->getValorMulta(),
			'valorDesconto' => $this->getValorDesconto(),
			'valorPago' => $this->getValorPago(),
			'historico' => $this->getHistorico(),
			'finStatusParcela' => $this->getFinStatusParcela(),
			'finTipoPagamento' => $this->getFinTipoPagamento(),
			'finChequeEmitido' => $this->getFinChequeEmitido(),
        ];
    }
}
